File: leetcode/problems/language-specific/go/219.contains-duplicate-ii.go
/*
 * @lc app=leetcode id=219 lang=golang
 *
 * [219] Contains Duplicate II
 */

// @lc code=start
func containsNearbyDuplicate(nums []int, k int) bool {
	if len(nums) <= 1 {
		return false
	}

	mappings := map[int]int{}
	for i, v := range nums {
		if mappings[v] > 0 && i-mappings[v] < k {
			return true
		}

		mappings[v] = i + 1
	}

	return false
}

// @lc code=end

File: leetcode/problems/language-specific/python/219.contains-duplicate-ii.py
#
# @lc app=leetcode id=219 lang=python3
#
# [219] Contains Duplicate II
#

# @lc code=start

# * Hash Table Solution | O(n) Time | O(n) Space
# * n -> The length of nums array


class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        nums_val_idx_map = collections.defaultdict(int)
        for idx, num in enumerate(nums):
            if num in nums_val_idx_map and abs(nums_val_idx_map[num] - idx) <= k:
                return True

            nums_val_idx_map[num] = idx

        return False


# @lc code=end
